### ISSUE TITLE
Against bitcoin p2p network sybil attackers.

### ISSUE THREAD

**Issue by chris-belcher** [chris-belcher/joinmarket#56]
Regarding https://www.reddit.com/r/Bitcoin/comments/2yvy6b/a_regulatory_compliance_service_is_sybil/

Obviously coinjoin doesn't directly help against this IP tracking. But this project can still help.

We should introduce a method where the taker can send the fully-signed tx hex to one of the makers who will then broadcast it. The taker would choose to broadcast the txhex himself, or send it to one randomly chosen maker to broadcast. In this way the coinjoin tx could come from many IPs instead of just the taker's.

Naturally this kind of bad actor could also be a sybil in the joinmarket. They will be limited in ways already discussed: To coinjoin everybody's transactions the sybil would need to own a large amount of bitcoins. Also their coinjoining activity will stop other sybils learning what's happening. Because of the feature in #14 the sybil will not be guaranteed to coinjoin with everyone just because they offer the lowest price.

It's a partial solution.

**Comment by AdamISZ** [comment]
> We should introduce a method where the taker (coinjoin-initiator) can send the fully-signed tx hex to one of the makers (market-makers who wait-to-coinjoin) who will then broadcast it.

Agreed, that can only help. Of course, network level obfuscation *might* be another step up, e.g. Tor, but that's another can of worms, and a bit outside the scope here.

**Comment by chris-belcher** [comment]
What does network level obfuscation mean in practice?

Regarding the edit. True.
The taker has to be in a race to claim the UTXO inputs, otherwise malicious takers could negotiate many coinjoins with makers and never broadcast them to use up all the maker's UTXOs.
The person controlling the IRC server will be easily able to make that link. Hopefully it shouldn't be an issue for another messaging channel like subspace.

**Comment by chris-belcher** [comment]
Wrote the code for the maker side. https://github.com/chris-belcher/joinmarket/commit/919e97fd2d58a9758f60fec350fd6888210945f9

Disabled on taker side for now.

**Comment by chris-belcher** [comment]
There could be a script called pushtx.py which connects to the network and sends a tx to a maker, either chosen by random or chosen by the user.

A similar kind of thing can be in the ob-watcher.py html page

**Comment by chris-belcher** [comment]
Could requesting a maker to push lots and lots of transactions could get that maker banned by other bitcoin nodes? This may be a DOS opportunity.

**Comment by chris-belcher** [comment]
An example of a paper that actually does this: https://arxiv.org/pdf/1612.06747v3.pdf

Nice paper with interesting results. They actually do spin up their own nodes which collect ip-address/transaction linkages. Then combined with the graph between bitcoin addresses they can figure out bitcoin-address/ip-address linkages.

Note that the data collection happened in late-2013, and the Bitcoin Core transaction relay algorithm has been changed significantly in the meantime to improve privacy (the assumptions in the introduction to the paper are therefore wrong today)
